py-polars/polars/internals/series/utils.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import annotations
+
+import inspect
+import sys
+from functools import wraps
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+import polars.internals as pli
+from polars.datatypes import DataType, dtype_to_ffiname
+
+if TYPE_CHECKING:
+    from polars.polars import PySeries
+
+    if sys.version_info >= (3, 10):
+        from typing import ParamSpec
+    else:
+        from typing_extensions import ParamSpec
+
+    T = TypeVar("T")
+    P = ParamSpec("P")
+    SeriesMethod = Callable[..., pli.Series]
+
+
+def _empty_() -> None:
+    """"""
+
+
+def _is_empty_method(func: SeriesMethod) -> bool:
+    """Confirm function ONLY has a docstring (eg: body is empty/pass)"""
+    ec, fc = _empty_.__code__, func.__code__
+    return (ec.co_code == fc.co_code) and (
+        len(fc.co_consts) == 2 and fc.co_consts[1] is None
+    )
```

Comment:
This function currently does not allow for an empty method with no docstring. I think it makes sense to support that, since I do see a future where we don't define docstrings for each version of the method, but re-use a single docstring.

Also, do we need the first check (`co_code`)? What does this do exactly? Isn't it enough to check those `co_const`?

And if we do need the check, is there no way to define this directly instead of first defining an empty function and using that?

Reply:
> I am still left wondering why PyCharm accepts this form of decorator but not the direct method decorator. Did you learn more about the nitty gritty here?

The class decorator works because static analysis (mypy/editor/etc) cannot know what that class decorator is going to do; so, static analysis sees the methods exactly as they are written in the code (undecorated - _until_ runtime). And at runtime, the `__signature__` injection in the decorator seems to help PyCharm finish the job (it gets _almost_ all the way there without it, but that was the missing piece of the puzzle to get proper autocomplete).

As for why it can't/doesn't follow the decorator properly and proxy the decorator's `__wrapped__` property signature instead (as VSCode does), I cannot say - let's just hope that eventually they fix that  :)

Reply:
Thanks for the explanation, and the extra commits! I think we're good to go now!